### ISSUE TITLE
Options to enable HTTP access to kernels

### ISSUE THREAD

**Issue by parente** [jupyter-incubator/kernel_gateway#29]
I've got it in my head that the kernel gateway can become a vehicle for providing many different ways of accessing kernels, both in terms of API and protocol. Right now, it supports HTTP to the notebook kernel CRUD API and Websockets carrying Jupyter protocol messages. We've talked about limiting channel access and doing HTTP to the notebook kernel CRUD API and SSE carrying Jupyter protocol iopub messages only.

I've got this (semi-crazy) idea that we could also optionally:

0. Turn off the websocket / kernel CRUD API
1. Have the kernel gateway read cell annotations from a notebook file on startup (option already supported)
2. Use those annotations to define HTTP endpoints referring to code cell to execute when they're invoked (new)
3. When one of those endpoints receives an HTTP request, have the kernel gateway send the request body/query args to the kernel (e.g., by executing them in an assignment statement, ...)
4. After sending the args, have the kernel gateway execute the cell code associated with the HTTP endpoint on the kernel and read its output (e.g., iopub stream message, ...)
5. Respond to the HTTP request with the output and/or error status appropriately

Pack the notebook + kernel gateway up in a container, scale that container image, and you've got a to use notebooks as a place to quickly prototype microservices that run on kernels.

**Comment by parente** [comment]
Whole bunch of brainstorming with @nitind and @Lull3rSkat3r on this today. Dumping the notes here. Nothing set in stone and we're totally open to getting feedback from others. (I think it'll be easier to discuss with the community when we have something working though to show the concept.)

### What's the tenancy / scalability of kernel microservices?

A kernel is single threaded and handles one request at a time. To scale, you scale the notebook+kernel gateway+kernel pattern at a container level.

-----------------------------

### What if the kernel is stateful across requests?

The burden of ensuring the API is stateless falls on the notebook author.

-----------------------------

### What's the start command?

`jupyter kernelgateway --KernelGatewayApp.api=kernel-http --KernelGatewayApp.ip=0.0.0.0 --KernelGatewayApp.seed_uri=my_annotated_notebook.ipynb --KernelGatewayApp.max_kernels=1 --KernelGateway.prespawn_count=1`

"Run the kernel gateway and immediately launch one kernel. Run the (unannotated cells?) contents of my_annotated_notebook.ipynb on that kernel when it launches. Expose a HTTP interface to the annotated cells in the my_annotated_notebook notebook."

Of course, we can simplify this over time.

-----------------------------

### What option do we need to add?

--KernelGatewayApp.api=kernel-http

Pick a better name / value pair. Need one for the default too, the existing Jupyter kernel/websocket API.

-------

### What's the annotation the notebook user puts in to specify the HTTP resource path?

Option 0: # <api:GET,/path/available/at>

***************** Option 1: # GET /path/:something/blah

Option 2: Use cell metadata instead (but then user is editing JSON or we have to write a toolbar or ...)

Pros of Non-UI
- Just works, no notebook server changes
- Jupyter guys seem to "laugh-at" squirreling metadata away in notebook

Cons
- Confusing, explicitly need to know what to do

------------------------------

### What do we want users to have to do to handle input/output in the notebook?

#### Option 0: Declare (if needed in your lang, e.g., Scala with "var") one variable that receives an entire request object JSON encoded

```
<cell>
REQUEST = '{}'
</cell>

<cell>
# POST /foo/<:something>
req = json.decode(REQUEST)
req['body']
req['args']['arg1']
req['path']['arg1']
req['host']
</cell>
```

#### ~~Option 1: Declare global BODY and ARGS and then json decode them~~

Pro: Flat notebook
Con: Have to set BODY / ARGS to default JSON for testing in notebook

#### ~~Option #2: Write everything as a function and json decode parameters~~

Pro: No extra work needed beyond writing a function
Con: Functions are rarer in notebooks

#### ~~Option #3: Annotate individual variables within an annotated cell and json decode them~~

Pro: No magic global requirement
Con: Have to define variables and remember to stick annotations on them

#### ~~Option #4: Include a mapping of args to local variables~~

```
<cell>
# Configure
# jsonEncode = json.encode
</cell>


<cell>
name = 'Talisker'
</cell>

<cell>
# GET /recommendations/<:scotch>
# :scotch -> name

#behind the scenes name = json.encode

get_recommendations(name)

</cell>
```

Pro: very explicit
Con: too explicit, leaks json handling to rest of notebook variable definition

------------------------------

### What's the mechanism for passing input to kernel / running the cell?

####  Option #0: Pass everything in ONE JSON blob

```
# PSEUDOCODE
cell = get_cell_for_http_endpoint(endpoint_name)
# package args, body, path params, etc. into a request dictionary in KG Python
kernel.execute('REQUEST ' + assignment_op_for_language(notebook.language) + json.encode(request))
kernel.execute(cell.source)
```

#### ~~Option #1: Pass as JSON~~

```
# PSEUDOCODE
cell = get_cell_for_http_endpoint(endpoint_name)
kernel.execute('BODY ' + assignment_op_for_language(notebook.language) + json.encode(request.body))
kernel.execute('ARGS ' + assignment_op_for_language(notebook.language) + json.encode(convert_to_dict(request.args))
kernel.execute(cell.source)
```

#### ~~Option #2: Invoke functions~~

```
# PSEUDOCODE
cell = get_cell_for_http_endpoint(endpoint_name)
function = parse_function_from_cell(cell.source)
kernel.execute(function_invocation_for_language(notebook.language, function, json.encode(request.body), json.encode(convert_to_dict(request.args))
```

#### ~~Option #3: Replace named args~~



```
cell = get_cell_for_http_endpoint(endpoint_name)
for named_arg in cell.source:
	kernel.execute(named_arg + assignment_op_for_language(notebook.language) + json.encode(request.args[named_arg]))
kernel.execute(cell.source)

<cell>
# ARG
scotch = 'Talisker'

get_recommendations(scotch)
</cell>
```

### What's the mechanism for getting output from kernel?

#### ~~Option 0: Put stuff in a RESPONSE (or back in REQUEST) in a known field~~

Pro: consitent with input
Con: means KG has to know how to read a dictionary in the language

#### ~~Options 1: Put a JSON blob in RESPONSE~~

Pro: consitent with input, makes it easy for KG to decode
Con: eveyrhing has to be a JSON blob as output

#### Option 2: Take the JSON envelope of the cell execution output and just return that

Pro: very easy, comes in the Jupyter protocol as-is
Con: Not a good REST API output for consumers of the API

#### Opt-in option 3: Make KG smarter about what media to pick out of response, maybe with extra annotations from the notebook

Improves upon the second option. Notebook author can opt-in but is not forced to do so.

```
<cell>
# GET /image
# Content-Type: image/png

plt.plot(something)
</cell>
```

### What's the "hello world" scenario?

We'll port the got scotch demo.

**Comment by parente** [comment]
"Hello world" edition to consider during development here: https://gist.github.com/parente/fe0e5992abe8946f67e9

**Comment by parente** [comment]
Prototype merged to master in #35. Opened #37 and #38 as follow-ons. jupyter-incubator/kernel_gateway_demos#1 and  jupyter-incubator/kernel_gateway_demos#2 will start to show how the feature can be put to use.
